projects/feray/feray/protocols.py

```diff
@@ -0,0 +1,124 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import (
+    Any,
+    Generic,
+    Iterable,
+    Protocol,
+    Union,
+)
+
+# Import all necessary types from the foundational data_structures module
+from .data_structures import (
+    DataFrameType,
+    DataVersion,
+    MerkleHash,
+    Provenance,
+    RowId,
+    S,
+    StorageId,  # <-- The newly added and crucial type
+)
```

Comment:
Curious to hear what you guys think about relative vs. absolute imports... I've personally pretty much banished relative imports from my code after a few bad experiences, and I also find it to be a bit more readable. But I realize this is also pretty down to opinions in a lot of cases, just curious to get yours.